ledger/voters.go

```diff
@@ -0,0 +1,337 @@
+		rewards := basics.PendingRewards(&ot, tr.Proto, acct.MicroAlgos, acct.RewardsBase, hdr.RewardsLevel)
+		money := ot.AddA(acct.MicroAlgos, rewards)
+		if ot.Overflowed {
+			return fmt.Errorf("votersTracker.loadTree: overflow adding rewards %d + %d", acct.MicroAlgos, rewards)
+		}
+
+		totalWeight = ot.AddA(totalWeight, money)
+		if ot.Overflowed {
+			return fmt.Errorf("votersTracker.loadTree: overflow computing totalWeight %d + %d", totalWeight.ToUint64(), money.ToUint64())
+		}
+
+		keyDilution := acct.VoteKeyDilution
+		if keyDilution == 0 {
+			keyDilution = tr.Proto.DefaultKeyDilution
+		}
+
+		participants[i] = compactcert.Participant{
+			PK:          acct.VoteID,
+			Weight:      money.ToUint64(),
+			KeyDilution: keyDilution,
+		}
+		addrToPos[acct.Address] = uint64(i)
+	}
+
+	tree, err := merklearray.Build(participants)
+	if err != nil {
+		return err
+	}
+
+	tr.mu.Lock()
+	tr.AddrToPos = addrToPos
+	tr.Participants = participants
+	tr.TotalWeight = totalWeight
+	tr.Tree = tree
+	tr.cond.Broadcast()
+	tr.mu.Unlock()
+
+	return nil
+}
+
+func (vt *votersTracker) newBlock(hdr bookkeeping.BlockHeader) {
+	proto := config.Consensus[hdr.CurrentProtocol]
+	if proto.CompactCertRounds == 0 {
+		// No compact certs.
+		return
+	}
+
+	// Check if any blocks can be forgotten because the compact cert is available.
+	for r, tr := range vt.round {
+		commitRound := r + basics.Round(tr.Proto.CompactCertVotersLookback)
+		certRound := commitRound + basics.Round(tr.Proto.CompactCertRounds)
+		if certRound <= hdr.CompactCertLastRound {
+			delete(vt.round, r)
+		}
+	}
+
+	// This might be a block where we snapshot the online participants,
+	// to eventually construct a merkle tree for commitment in a later
+	// block.
+	r := uint64(hdr.Round)
+	if (r+proto.CompactCertVotersLookback)%proto.CompactCertRounds == 0 {
+		_, ok := vt.round[basics.Round(r)]
+		if ok {
+			vt.au.log.Errorf("votersTracker.newBlock: round %d already present", r)
+		} else {
+			err := vt.loadTree(hdr)
+			if err != nil {
+				vt.au.log.Warnf("votersTracker.newBlock: loadTree: %v", err)
+			}
+		}
+	}
+}
+
+// lowestRound() returns the lowest round state (blocks and accounts) needed by
+// the votersTracker in case of a restart.  The accountUpdates tracker will
+// not delete account state before this round, so that after a restart, it's
+// possible to reconstruct the votersTracker.  If votersTracker does
+// not need any blocks, it returns base.
+func (vt *votersTracker) lowestRound(base basics.Round) basics.Round {
+	minRound := base
+	for r := range vt.round {
+		if r < minRound {
+			minRound = r
+		}
+	}
+	return minRound
+}
+
+// getVoters() returns the top online participants from round r.
+func (vt *votersTracker) getVoters(r basics.Round) (*VotersForRound, error) {
+	tr, ok := vt.round[r]
+	if !ok {
+		// Not tracked: compact certs not enabled.
+		return nil, nil
+	}
+
+	// Wait for the Merkle tree to be constructed.
+	tr.mu.Lock()
+	for tr.Tree == nil {
```

Comment:
this could lead to a hang in case we had error during the `VotersForRound::loadTree`. maybe store the error generated during the `VotersForRound::loadTree` inside the `VotersForRound`, and make sure to broadcast when changing the error as well ?

Reply:
That seems like a good plan.  I think originally I had no failure paths in the background `loadTree()` function, but indeed now there's possible errors that can cause this to hang forever.  Pushed a revised version.